invariance.md

```diff
@@ -0,0 +1,204 @@
+# What is Invariance?
+
+The term 'invariant' is one of the linchpins of Decision Making Isolation. It is used in the [Cpp Core Guidelines](https://isocpp.github.io/CppCoreGuidelines) as the reason to use a class versus struct. But what does it really mean?
+
+## Defining Invariance
+
+In [C.2](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c2-use-class-if-the-class-has-an-invariant-use-struct-if-the-data-members-can-vary-independently), the Guidelines define invariance as:
+
+> An invariant is a logical condition for the members of an object that a constructor must establish for the public member functions to assume. After the invariant is established (typically by a constructor) every member function can be called for the object. An invariant can be stated informally (e.g., in a comment) or more formally using Expects.
+>
+> If all data members can vary independently of each other, no invariant is possible.
+
+I've read this explanation over and over, but it still never quite made sense in those terms. Let's propose an alternative definition of invariance:
+
+> [!Important]  
+> An invariant is a logical condition for the members of an object that can be expected to be true from the time the constructor exits to the time the destructor is called. These logical conditions fall into three categories.
+>
+> * A relationship between at least two data members such that all members cannot vary independently.
+> * A relationship between a data member and a resource (aka RAII).
+> * A property of a data member that must always be true.
+
+Let's break down each of these.
+
+## Relationships Between Members
+
+Here's a simple example.
+
+Consider a simple electrical circuit that is characterized by Ohm's Law:
+
+> volts = current * resistance
+
+Ohm's law dictates that if you vary any one of the parameters, at least one other must also change. If the voltage increases, either the current or resistance must also increase. If the current decreases, either the resistance will increase or the voltage will decrease. They cannot be changed independently. If we were to code up an object that held these parameters, the parameters couldn't be public members. We'd need to enforce the invariance dictated by Ohm's Law.
```

Comment:
You bring up that the parameters shouldn't be public members here, which is true, but also it took me a bit to get that your unspoken point was that they shouldn't be in a struct. Because the topic is "invariance", and it was a take off from the class vs struct, I wasn't in the headspace to be reflecting on this as an angle.

Reply:
I added a section at the beginning that aims to clarify the goal in understanding invariance.